### ISSUE TITLE
Implement SqlDataSource

### ISSUE THREAD

**Issue by roji** [dotnet/SqlClient#2119]
.NET 7.0 introduced the new DbDataSource abstraction to System.Data ([issue](https://github.com/dotnet/runtime/issues/64812)), [here's a video](https://www.youtube.com/watch?v=vRUtHeUpU44&list=PLdo4fOcmZ0oX-DBuRG4u58ZTAJgBAeQ-t&index=71) discussing it in depth.

Now, the .NET 7.0 implementation provides a default implementation which can be obtained from the ADO.NET provider's DbProviderFactory without the driver needing to implement anything. So the following code already works today:

```c#
await using var dataSource = SqlClientFactory.Instance.CreateDataSource("<connection string>");
await using var connection = await dataSource.OpenConnectionAsync();
await using var command = connection.CreateCommand();
command.CommandText = "SELECT 1";
Console.WriteLine(command.ExecuteScalar());
```

However, this provides a database-agnostic DbDataSource which isn't typed to any specific provider, so it hands out DbConnections rather than SqlConnections, which may be cumbersome to work with if the user is only targeting SQL Server.

A  proper SqlDataSource implementation in SqlClient would solve this, and also provide a good configuration/setup point (e.g. for authentication/access tokens, rather than the current static methods).

**Comment by JRahnama** [comment]
Thanks @roji. We will discuss this internally and will get back to you.

**Comment by ErikEJ** [comment]
@roji Are you thinking of something like this?

```
builder.Services.AddSqlDataSource("Server=.\\SQLEXPRESS;Database=Northwind;Integrated Security=true;Trust Server Certificate=true", setupAction =>
   {
       setupAction.UseAccessToken(accessToken);
   });

```

**Comment by roji** [comment]
Yep, something like that. Npgsql has an NpgsqlDataSourceBuilder type for configuring an NpgsqlDataSource (when you're done you call Build() on it and get your data source); that would be the type of the `setupAction` parameter in your lambda above. There's also an [Npgsql.DependencyInjection](https://www.nuget.org/packages/Npgsql.DependencyInjection#readme-body-tab) package which does precisely this kind of nice integration with DI that you're showing above.

NpgsqlDataSourceBuilder has all manner of configuration on it: stuff for managing access tokens, tweaking type mappings, logging and telemetry configuration, etc. Basically anything that you need to pre-configure for all the connections you'll be using from that data source. This is a much better alternative than static configuration, which is what you have to do if there's no data source and users are directly instantiating SqlConnections.

(see the video I posted above for more discussion on how this is useful etc.)

**Comment by roji** [comment]
Note that as a first step, I'd just implement SqlDataSource so that users can use it as a simple connection factory (e.g. to be registered in DI). Adding actual support for e.g. access token management is additional work on top of that which would require a bit more refactoring inside SqlClient to support.

**Comment by ErikEJ** [comment]
Cool, I might look into that, since I already published a package which does that :smile:

**Comment by ErikEJ** [comment]
@roji I presume then something like ErikEJ.Data.SqlClient.DependencyInjection would be next (and maybe eventually Microsoft.Data.SqlClient.DependencyInjection) ?

**Comment by roji** [comment]
Yeah, exactly. SqlDataSource is something that needs to be implemented inside SqlClient (since ideally it would go on to support actual SqlClient-specific configurations like around access tokens). A DI integration package - similar to Npgsql.DependencyInjection - could be external (e.g. ErikEJ...), though ideally it would indeed be Microsoft.Data.SqlClient.DependencyInjection at some point.

**Comment by adamsitnik** [comment]
Once it gets implemented, AspNetCore.HealthChecks.SqlServer (a very popular health check package for SQL Server with 37 milion downloads) should start using it too. I've created https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/issues/2123 to track it.

**Comment by wasabii** [comment]
I have to ping on this one. What is the status? I am writing an application that would be dependant upon this, but I don't see much of a concrete update since over a year ago. The code does seem to exist in a branch.

**Comment by roji** [comment]
@wasabii no application should be **dependent** on this - you should always be able to write your application without SqlDataSource (as applications have been done for many years).
